virtcontainers/cgroups.go

```diff
@@ -0,0 +1,192 @@
+// Copyright (c) 2018 Huawei Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+package virtcontainers
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/containerd/cgroups"
+	"github.com/kata-containers/runtime/virtcontainers/pkg/annotations"
+	specs "github.com/opencontainers/runtime-spec/specs-go"
+)
+
+const (
+	vcpuGroupName       = "vcpu"
+	defaultCgroupParent = "/vc"
+)
+
+type sandboxCgroups struct {
+	commonParent cgroups.Cgroup
+	sandboxSub   cgroups.Cgroup
+	vcpuSub      cgroups.Cgroup
+}
+
+func (s *Sandbox) newCgroups() error {
+	// New will still succeed when cgroup exists
+	// create common parent for all kata-containers
+	// e.g. /sys/fs/cgroup/cpu/vc
+	parent, err := cgroups.New(cgroups.V1,
```

Comment:
cgroup should be created in the path specified by `cgroupsPath` in the `config.json`, Q. why you are not using it?

Reply:
@WeiZhang555 @devimc  maybe we should consider  `docker cli` `Cgroup Parent` option.  ^^

Reply:
Considering one valid use case though we don't recommend it at all:
one POD with 3 containers: c1, c2, c3

c1 has cgroups path: `/mycontainer/c1`
c2 has cgroups path: `/hello/c2`
c3 has cgroups path: `/world/c3`

So how to find a shared parent for 3 containers to put the qemu in?

Reply:
@devimc I know we were seeing some file descriptor leaks when ignoring `cgroupsPath` in the `config.json` back in the day. Do we have tests in the CI for detecting that? I dont recall the root-cause, but do you think this implementation should still take care of that?

Reply:
There're no CI tests for this yet. In normal cases, `destroyCgroups` will ensure all cgroup files being deleted if this function can be invoked.

Reply:
Reading https://github.com/opencontainers/runtime-spec/blob/master/config-linux.md#control-groups, we do have to honour `cgroupsPath`.

But I'm a little confused here...

Docker doesn't have the concept of pod/sandbox so we create one VM instance per container.

- If the user runs "`docker run ...`" Docker will create a `config.json` specifying a cgroup of `/docker/${cid}/` we can create `/sys/fs/cgroup/docker/${cid}/` for the hypervisor and the shim.

- If the user then runs "`docker exec -ti ${cid} sh`", the second shim process we create should also be added to the same cgroup (which doesn't appear to be happening today!?!)

In other words, why does it matter than Docker doesn't have the concept of pods? K8s does, but then it wouldn't be generating a `cgroupsPath = "/docker/${cid}/`" ...?

Reply:
I don't know what are going to be the consequences of whether `cgroupsPath` is ignored, what I'm sure is that systemd-cgroups won't be honoured and cgroup can't be updated with `systemctl`

Reply:
@jodh-intel I do want to honor `cgroupsPath` but I didn't find a reasonable logic for handling this.

> Docker doesn't have the concept of pod/sandbox so we create one VM instance per container.

That's not 100% true. We do want to run K8S+docker, so we should still regard docker as one VM instance per POD.

>In other words, why does it matter than Docker doesn't have the concept of pods? K8s does, but then it wouldn't be generating a cgroupsPath = "/docker/${cid}/" ...?

I can't find a common logic for handling this, consider a "best" situation:

```
c1 has cgroups path: /POD1/c1
c2 has cgroups path: /POD1/c2
c3 has cgroups path: /POD1/c3
```

According to this configuration, we can construct a cgroup hierarchy as:

```
/sys/fs/cgroup/cpu
|--POD1
     |--vcpu
     |--c1
     |--c2
     |--c3
```

we can put VM process in `vcpu` cgroup to limit the qemu processes, and put shim process of container c1/c2/c3 into cgroup c1/c2/c3, this makes lots of sense and makes a perfect logic.

But take the worst situation as this:
```
c1 has cgroups path: /mycontainer/c1
c2 has cgroups path: /hello/c2
c3 has cgroups path: /world/c3
```

Then how to hornor the `cgroupPath` and how to find a good common parent? What I can think of is we only honor cgroupPath of `c1` and ignores c2 and c3.

```
/sys/fs/cgroup/cpu
|--mycontainer
     |--vcpu
     |--c1
     |--c2
     |--c3
```

I'm not confident enough that this makes sense or not, and as we create the container of each pod one by one, we can't control how user will define the cgroupsPath for each container.